File: apps/macos/ShibaStudio/AppIdentity.swift
import Foundation

enum AppIdentity {
    static let preferredPort = 18765
    static let manifestURL = URL(string: "https://shiba-studio.io/packages/manifest.json")!
    static let packagesPage = URL(string: "https://shiba-studio.io/packages.html")!
    static let productName = "Shiba Studio"

    static var supportDirectory: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/ShibaStudio")
    }

    static var dataDirectory: URL { supportDirectory.appendingPathComponent("data") }
    static var prefsFile: URL { supportDirectory.appendingPathComponent("prefs.json") }
    static var updatesDirectory: URL { supportDirectory.appendingPathComponent("updates") }
    static var logFile: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Logs/ShibaStudio/studio.log")
    }

    static var runtimeDirectory: URL {
        if let resources = Bundle.main.resourceURL {
            let bundled = resources.appendingPathComponent("runtime")
            if FileManager.default.fileExists(atPath: bundled.path) {
                return bundled
            }
        }
        return Bundle.main.bundleURL.appendingPathComponent("Contents/Resources/runtime")
    }

    static var appJSON: URL { runtimeDirectory.appendingPathComponent("app.json") }
    static var nodeBinary: URL { runtimeDirectory.appendingPathComponent("bin/node") }
    static var nextCLI: URL { runtimeDirectory.appendingPathComponent("node_modules/next/dist/bin/next") }

    static func ensureUserFolders() {
        let folders = [dataDirectory, updatesDirectory, logFile.deletingLastPathComponent()]
        for folder in folders {
            try? FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        }
    }

    static func isBundledRuntime() -> Bool {
        let fm = FileManager.default
        return fm.fileExists(atPath: runtimeDirectory.appendingPathComponent("package.json").path)
            && fm.fileExists(atPath: runtimeDirectory.appendingPathComponent(".next").path)
            && fm.isExecutableFile(atPath: nodeBinary.path)
            && fm.fileExists(atPath: nextCLI.path)
    }

    static func readStamp() -> Stamp {
        guard let data = try? Data(contentsOf: appJSON),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return Stamp(channel: "development", sha: nil, builtAt: nil)
        }
        return Stamp(
            channel: object["channel"] as? String,
            sha: object["sha"] as? String,
            builtAt: object["builtAt"] as? String
        )
    }

    static func readPrefs() -> Prefs {
        guard let data = try? Data(contentsOf: prefsFile),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return Prefs(channel: nil, autoUpdate: true, keepInMenuBar: true)
        }
        return Prefs(
            channel: object["channel"] as? String,
            autoUpdate: (object["autoUpdate"] as? Bool) ?? true,
            keepInMenuBar: (object["keepInMenuBar"] as? Bool) ?? true
        )
    }

    static func writePrefs(_ prefs: Prefs) {
        ensureUserFolders()
        var object: [String: Any] = [
            "autoUpdate": prefs.autoUpdate,
            "keepInMenuBar": prefs.keepInMenuBar,
        ]
        if let channel = prefs.channel { object["channel"] = channel }
        if let data = try? JSONSerialization.data(withJSONObject: object, options: [.prettyPrinted]) {
            try? data.write(to: prefsFile)
        }
    }

    static func resolvedChannel() -> String {
        if let prefs = readPrefs().channel, prefs == "main" || prefs == "development" {
            return prefs
        }
        if let stamp = readStamp().channel, stamp == "main" || stamp == "development" {
            return stamp
        }
        return "development"
    }

    static func shortSHA() -> String {
        let sha = readStamp().sha ?? ""
        return sha.count <= 7 ? sha : String(sha.prefix(7))
    }

    struct Stamp {
        let channel: String?
        let sha: String?
        let builtAt: String?
    }

    struct Prefs {
        var channel: String?
        var autoUpdate: Bool
        var keepInMenuBar: Bool
    }
}

File: apps/windows/AppIdentity.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ShibaStudio;

sealed record AppStamp(
    int Version,
    string? Kind,
    string? Platform,
    string? Channel,
    string? Sha,
    string? BuiltAt,
    string? Node,
    int? PreferredPort,
    string? ManifestUrl,
    string? PackagesPage);

sealed class UserPrefs
{
    public string? Channel { get; set; }
    public bool AutoUpdate { get; set; } = true;
    /// <summary>Minimize and close hide the window to the notification area instead of exiting.</summary>
    public bool MinimizeToTray { get; set; } = true;
}

static class AppIdentity
{
    public const int PreferredPort = 18765;
    public const string ManifestUrl = "https://shiba-studio.io/packages/manifest.json";
    public const string PackagesPage = "https://shiba-studio.io/packages.html";
    public const string ProductName = "Shiba Studio";

    static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public static string InstallDirectory => AppContext.BaseDirectory.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);

    public static string UserDirectory => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "ShibaStudio");

    public static string DataDirectory => Path.Combine(UserDirectory, "data");
    public static string WebViewDirectory => Path.Combine(UserDirectory, "webview");
    public static string LogDirectory => Path.Combine(UserDirectory, "logs");
    public static string LogFile => Path.Combine(LogDirectory, "studio.log");
    public static string PrefsFile => Path.Combine(UserDirectory, "prefs.json");
    public static string UpdatesDirectory => Path.Combine(UserDirectory, "updates");

    public static string RuntimeDirectory
    {
        get
        {
            var bundled = Path.Combine(InstallDirectory, "runtime");
            if (Directory.Exists(bundled)) return bundled;
            return InstallDirectory;
        }
    }

    public static string AppJsonPath => Path.Combine(RuntimeDirectory, "app.json");

    public static string NodeBinary
    {
        get
        {
            var bundled = Path.Combine(RuntimeDirectory, "node.exe");
            if (File.Exists(bundled)) return bundled;
            bundled = Path.Combine(RuntimeDirectory, "bin", "node.exe");
            if (File.Exists(bundled)) return bundled;
            return bundled;
        }
    }

    public static string NextCli => Path.Combine(RuntimeDirectory, "node_modules", "next", "dist", "bin", "next");

    public static bool IsBundledRuntime()
    {
        return File.Exists(Path.Combine(RuntimeDirectory, "package.json"))
            && Directory.Exists(Path.Combine(RuntimeDirectory, ".next"))
            && File.Exists(NextCli)
            && File.Exists(NodeBinary);
    }

    public static AppStamp ReadStamp()
    {
        if (!File.Exists(AppJsonPath))
        {
            return new AppStamp(1, "bundled-desktop", "windows", "development", null, null, null, PreferredPort, ManifestUrl, PackagesPage);
        }
        try
        {
            var stamp = JsonSerializer.Deserialize<AppStamp>(File.ReadAllText(AppJsonPath), JsonOptions);
            return stamp ?? new AppStamp(1, "bundled-desktop", "windows", "development", null, null, null, PreferredPort, ManifestUrl, PackagesPage);
        }
        catch (JsonException)
        {
            return new AppStamp(1, "bundled-desktop", "windows", "development", null, null, null, PreferredPort, ManifestUrl, PackagesPage);
        }
    }

    public static UserPrefs ReadPrefs()
    {
        if (!File.Exists(PrefsFile)) return new UserPrefs();
        try
        {
            return JsonSerializer.Deserialize<UserPrefs>(File.ReadAllText(PrefsFile), JsonOptions) ?? new UserPrefs();
        }
        catch (JsonException)
        {
            return new UserPrefs();
        }
    }

    public static void WritePrefs(UserPrefs prefs)
    {
        Directory.CreateDirectory(UserDirectory);
        File.WriteAllText(PrefsFile, JsonSerializer.Serialize(prefs, JsonOptions));
    }

    public static string ResolvedChannel()
    {
        var prefs = ReadPrefs().Channel?.Trim();
        if (prefs is "main" or "development") return prefs;
        var stamp = ReadStamp().Channel?.Trim();
        return stamp is "main" or "development" ? stamp : "development";
    }

    public static string ShortSha()
    {
        var sha = ReadStamp().Sha?.Trim() ?? "";
        return sha.Length <= 7 ? sha : sha[..7];
    }

    public static void EnsureUserFolders()
    {
        Directory.CreateDirectory(DataDirectory);
        Directory.CreateDirectory(WebViewDirectory);
        Directory.CreateDirectory(LogDirectory);
        Directory.CreateDirectory(UpdatesDirectory);
    }
}
